python/docs/basics.md

```diff
@@ -0,0 +1,134 @@
+# RMM - the RAPIDS Memory Manager
+
+There are two ways to use RMM in Python code:
+
+1. Using the `rmm.DeviceBuffer` API to explicitly create and manage
+   device memory allocations
+2. Transparently via external libraries such as CuPy and Numba
+
+RMM provides a `MemoryResource` abstraction to control _how_ device
+memory is allocated in both the above uses.
+
+### DeviceBuffers
+
+A DeviceBuffer represents an **untyped, uninitialized device memory
+allocation**.  DeviceBuffers can be created by providing the
+size of the allocation in bytes:
+
+```python
+>>> import rmm
+>>> buf = rmm.DeviceBuffer(size=100)
+```
+
+The size of the allocation and the memory address associated with it
+can be accessed via the `.size` and `.ptr` attributes respectively:
+
+```python
+>>> buf.size
+100
+>>> buf.ptr
+140202544726016
+```
+
+DeviceBuffers can also be created by copying data from host memory:
+
+```python
+>>> import rmm
+>>> import numpy as np
+>>> a = np.array([1, 2, 3], dtype='float64')
+>>> buf = rmm.to_device(a.tobytes())
+>>> buf.size
+24
+```
+
+Conversely, the data underlying a DeviceBuffer can be copied to the
+host:
+
+```python
+>>> np.frombuffer(buf.tobytes())
+array([1., 2., 3.])
+```
+
+### MemoryResource objects
+
+`MemoryResource` objects are used to configure how device memory allocations are made by
+RMM.
+
+By default if a `MemoryResource` is not set explicitly, RMM uses the `CudaMemoryResource`, which 
+uses `cudaMalloc` for allocating device memory.
+
+`rmm.reinitialize()` provides an easy way to initialize RMM with specific memory resource options
+across multiple devices. See `help(rmm.reinitialize)` for full details.
+
+For lower-level control, the `rmm.mr.set_current_device_resource()` function can be
+used to set a different MemoryResource for the current CUDA device.  For
+example, enabling the `ManagedMemoryResource` tells RMM to use
+`cudaMallocManaged` instead of `cudaMalloc` for allocating memory:
+
+```python
+>>> import rmm
+>>> rmm.mr.set_current_device_resource(rmm.mr.ManagedMemoryResource())
+```
+
+> :warning: The default resource must be set for any device **before**
+> allocating any device memory on that device.  Setting or changing the
+> resource after device allocations have been made can lead to unexpected
+> behaviour or crashes. See [Multiple Devices](#multiple-devices)
+
+As another example, `PoolMemoryResource` allows you to allocate a
+large "pool" of device memory up-front. Subsequent allocations will
+draw from this pool of already allocated memory.  The example
+below shows how to construct a PoolMemoryResource with an initial size
+of 1 GiB and a maximum size of 4 GiB. The pool uses
+`CudaMemoryResource` as its underlying ("upstream") memory resource:
+
+```python
+>>> import rmm
+>>> pool = rmm.mr.PoolMemoryResource(
+...     upstream=rmm.mr.CudaMemoryResource(),
+...     initial_pool_size=2**30,
+...     maximum_pool_size=2**32
+... )
+>>> rmm.mr.set_current_device_resource(pool)
+```
```

Comment:
It may be good to show an example of creating a Pool of managed memory as well.

Reply:
writ